PWGCF/GenericFramework/Tasks/flowGenericFramework.cxx

```diff
@@ -390,58 +381,64 @@ struct GenericFramework {
-      QAregistry.fill(HIST("phi"), phi);
-      QAregistry.fill(HIST("eta"), eta);
-      QAregistry.fill(HIST("pt"), pt);
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), mcParticle.phi(), mcParticle.eta(), vtxz, wacc);
+      if (cfgFillQA)
+        FillQA<kData>(mcParticle.phi(), mcParticle.eta(), mcParticle.pt(), track.dcaXY(), track.dcaZ());
+      FillGFW(mcParticle.phi(), mcParticle.eta(), mcParticle.pt(), weff * wacc);
     } else if constexpr (framework::has_type_v<aod::mcparticle::McCollisionId, typename TrackObject::all_columns>) {
-      pt = track.pt();
-      eta = track.eta();
-      phi = track.phi();
-      if (!track.isPhysicalPrimary() || eta < etalow || eta > etaup || pt < ptlow || pt > ptup)
+      if (!track.isPhysicalPrimary() || track.eta() < etalow || track.eta() > etaup || track.pt() < ptlow || track.pt() > ptup)
         return;
-      if (cfgFillQA) {
-        QAregistry.fill(HIST("phi_gen"), phi);
-        QAregistry.fill(HIST("eta_gen"), eta);
-        QAregistry.fill(HIST("pt_gen"), pt);
-      }
+      if (cfgFillQA)
+        FillQA<kGen>(track.phi(), track.eta(), track.pt());
+      FillGFW(track.phi(), track.eta(), track.pt(), 1.);
     } else {
-      pt = track.pt();
-      eta = track.eta();
-      phi = track.phi();
       if (cfgFillWeights)
-        fWeights->Fill(phi, eta, vtxz, pt, centrality, 0);
-      if (!setCurrentParticleWeights(weff, wacc, phi, eta, pt, vtxz))
+        fWeights->Fill(track.phi(), track.eta(), vtxz, track.pt(), centrality, 0);
+      if (!setCurrentParticleWeights(weff, wacc, track.phi(), track.eta(), track.pt(), vtxz))
         return;
-      registry.fill(HIST("phi_eta_vtxZ_corrected"), phi, eta, vtxz, wacc);
-      if (cfgFillQA) {
-        QAregistry.fill(HIST("phi"), phi);
-        QAregistry.fill(HIST("eta"), eta);
-        QAregistry.fill(HIST("pt_dcaXY_dcaZ"), pt, track.dcaXY(), track.dcaZ());
-      }
+      registry.fill(HIST("phi_eta_vtxZ_corrected"), track.phi(), track.eta(), vtxz, wacc);
+      if (cfgFillQA)
+        FillQA<kData>(track.phi(), track.eta(), track.pt(), track.dcaXY(), track.dcaZ());
+      FillGFW(track.phi(), track.eta(), track.pt(), weff * wacc);
     }
+  }
+
+  template <typename T>
+  inline void FillGFW(T phi, T eta, T pt, float w)
+  {
     bool WithinPtPOI = (ptpoilow < pt) && (pt < ptpoiup); // within POI pT range
     bool WithinPtRef = (ptreflow < pt) && (pt < ptrefup); // within RF pT range
     if (WithinPtRef)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 1);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 1);
     if (WithinPtPOI)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 2);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 2);
     if (WithinPtPOI && WithinPtRef)
-      fGFW->Fill(track.eta(), fPtAxis->FindBin(pt) - 1, phi, wacc * weff, 4);
+      fGFW->Fill(eta, fPtAxis->FindBin(pt) - 1, phi, w, 4);
+    return;
+  }
```

Comment:
For the next iteration consider templating on the the track/particle object instead
Both have the same getters and your code will be simpler and more efficient